### DISCUSSION TITLE
Is it possible to maintain tree-shaking an object when accessing a non-existing property?

### DISCUSSION THREAD

**Discussion by FunctionDJ** [webpack/webpack#15245] [Q&A] [answered]
Suppose we have a module like this:
```js
export default {
  foo: "",
  bar: ""
}
```
This is how the module is used:
```js
import values from "./module";
console.log(values.doesNotExist);
```
This will cause webpack to add the entire upper module into the bundle, probably to protect any potential black-magic side-effects that could make this property access work.
Is it possible to disable this fallback and keep tree-shaking enabled?

**Comment by alexander-akait** [comment, marked as answer]
I am afraid not, but we have https://github.com/webpack/webpack/issues/14814? But you should get an warning on it when your bundle, do you real have ESM code, maybe you have babel?

**Reply by FunctionDJ** [reply]
@alexander-akait Some more context: The upper bundle code gets generated by a custom loader, so we don't know the available object properties beforehand. Ideally we'd use named exports, but TypeScript doesn't support this, it must know the available exports beforehand.
A default export is the only way we can use the type `Record<string, string>`.

**Reply by FunctionDJ** [reply]
I checked #14814 but unfortunately doesn't seem to apply in this case :/

**Reply by alexander-akait** [reply]
I am afraid it is very specific so we don't support it (and it is exotic case), you can try to create plugin and implement this, but it requires more time and knowledge about webpack

**Reply by vankop** [reply]
I think you can mark module as `javascript/dynamic` via webpack.config https://webpack.js.org/configuration/module/#ruletype

> Is it possible to disable this fallback and keep tree-shaking enabled?

but this will not be fixed..
